src/java.base/share/classes/java/lang/foreign/Linker.java

```diff
@@ -197,16 +202,29 @@
  * </tbody>
  * </table></blockquote>
  * <p>
- * All the native linker implementations limit the function descriptors that they support to those that contain
- * only so-called <em>canonical</em> layouts. A canonical layout has the following characteristics:
+ * More formally, all native linker implementations can only operate on a subset of memory layouts, called <em>supported layouts</em>.
+ * A layout {@code L} is supported iff:
+ * <ul>
+ * <li>{@code L} is a value layout {@code V} and {@code V.withoutName()} is a canonical layout</li>
+ * <li>{@code L} is an address layout {@code A} and {@code A.withoutTargetLayout().withoutName()} is a canonical layout</li>
+ * <li>{@code L} is a sequence layout {@code S} and all the following conditions hold:
+ * <ol>
+ * <li>the alignment constraint of {@code S} is set to its <a href="MemoryLayout.html#layout-align">natural alignment</a>, and</li>
+ * <li>{@code S.elementLayout()} is a supported layout.</li>
+ * </ol>
+ * </li>
+ * <li>{@code L} is a group layout {@code G} and all the following conditions hold:
  * <ol>
- * <li>Its alignment constraint is set to its <a href="MemoryLayout.html#layout-align">natural alignment</a></li>
- * <li>If it is a {@linkplain ValueLayout value layout}, its {@linkplain ValueLayout#order() byte order} is
- * the {@linkplain ByteOrder#nativeOrder() native byte order}.
- * <li>If it is a {@linkplain GroupLayout group layout}, its size is a multiple of its alignment constraint, and</li>
- * <li>It does not contain padding other than what is strictly required to align its non-padding layout elements,
- * or to satisfy constraint 3</li>
+ * <li>the alignment constraint of {@code G} is set to its <a href="MemoryLayout.html#layout-align">natural alignment</a>;</li>
+ * <li>the size of {@code G} is a multiple of its alignment constraint;</li>
```

Comment:
Do you think this is a constraint that will hold across all linker implementations?

Reply:
All "native linkers" as the text says. Other linkers (e.g. not for C) might obey other rules. These rules are basically constraining  struct layouts to what can come out of a native compiler in the absence of pragma pack directives.

Reply:
Ah, i see now, i missed the relevance of that term, introduced earlier in non-modified text.